bootstrap.py

```diff
@@ -0,0 +1,211 @@
+
+    if verbose:
+        print(f'Push: {push_command}')
+
+    if dry_run:
+        return True
+
+    if not subprocess.run(push_command, stderr=None, check=False).returncode:
+        return True
+
+    print(f'Failed to push image {tag}.', file=sys.stderr)
+    return False
+
+
+def tag_image(tag, verbose=False, dry_run=False):        # pylint: disable=redefined-outer-name
+    """Tag an existing 'dunes' image with the provided tag."""
+
+    if tag is None:
+        return False
+
+    tag_command = ['docker', 'tag', 'dunes', tag]
+
+    if verbose:
+        print(f'Tagging: {tag_command}')
+
+    if dry_run is True:
+        return True
+
+    # expect a zero result for success
+    if not subprocess.run(tag_command, stderr=None, check=False).returncode:
+        return True
+
+    print(f'Failed to apply tag: {tag}', file=sys.stderr)
+    return False
+
+
+def build_image(leap_version=None, cdt_version=None, refcon_version=None, tag='dunes', verbose=False, dry_run=False):  # pylint: disable=too-many-arguments, redefined-outer-name
+    """Build a docker image for working with DUNES.
+
+    """
+
+    barg = '--build-arg'
+
+    # Set user id to a default value
+    user_id = 0
+
+    # Set group id based on platform
+    group_id = 0
+    if platform.system() == "Darwin":
+        group_id = 200
+
+    # pylint: disable=line-too-long
+    build_command = ['docker', 'build', '--no-cache', '-f','Dockerfile', '-t',tag, barg,f'USER_ID={user_id}', barg,f'GROUP_ID={group_id}', DUNES_ROOT]
+
+    if leap_version:
+        build_command.extend( [barg, f'LEAP_VERSION={leap_version}'] )
+
+    if cdt_version:
+        build_command.extend( [barg, f'CDT_VERSION={cdt_version}'] )
+
+    if refcon_version:
+        build_command.extend( [barg, f'REFCON_VERSION={refcon_version}'] )
+
+    if verbose:
+        print(f'Build command: {build_command}' )
+
+    if dry_run:
+        return True
+
+    if not subprocess.run(build_command, stderr=None, check=False).returncode:
+        return True
+
+    print('Failed to build the image.', file=sys.stderr)
+    return False
+
+
+if __name__ == "__main__":
+
+    # Create the CLI code
+    parser = argparse.ArgumentParser(
+        prog='bootstrap',
+        description='Generate a Docker image for DUNES.'
+        #epilog='Text at the bottom of help'
+    )
+
+    # Add arguments.
+    #  pylint:disable=line-too-long
+    parser.add_argument('-c', '--cdt',    dest='cdt_version',    action='store',      help='Set the leap version to CDT_VERSION.')
+    parser.add_argument('-l', '--leap',   dest='leap_version',   action='store',      help='Set the CDT version to LEAP_VERSION.')
+    parser.add_argument('--refcon',       dest='refcon_version', action='store',      help='Set the reference-contract version to REFCON_VERSION. Usually a commit hash.')
+    parser.add_argument('-r','--release', dest='release_build',  action='store_true', help='Tag the image with the version reported by DUNES.')
```

Comment:
I tried command `.\bootstrap.bat --refcon main -r`
and I it seems it did not get tagged properly, because at the end I got:
```
(...)
Removing intermediate container 57f931841149
 ---> b9fa020aeb78
(...)
Successfully built b9fa020aeb78
Successfully tagged dunes:latest
SECURITY WARNING: You are building a Docker image from Windows against a non-Windows Docker host. All files and directories added to build context will have '-rwxr-xr-x' permissions. It is recommended to double check and reset permissions for sensitive files and directories.
PS C:\Users\mikel\repo\DUNE>
PS C:\Users\mikel\repo\DUNE> docker images
REPOSITORY   TAG       IMAGE ID       CREATED          SIZE
dunes        732c0ac   b9fa020aeb78   19 seconds ago   3.23GB
dunes        latest    b9fa020aeb78   19 seconds ago   3.23GB
<none>       <none>    d85260086c61   8 minutes ago    72.8MB
<none>       <none>    3f6acc318e17   16 minutes ago   3.17GB
<none>       <none>    0cd6467ecac3   25 minutes ago   3.24GB
<none>       <none>    7750791f5714   34 minutes ago   3.23GB
<none>       <none>    81b6362b7970   2 hours ago      72.8MB
<none>       <none>    255ba588fd8b   3 hours ago      72.8MB
<none>       <none>    5bb0649c831d   3 hours ago      72.8MB
ubuntu       latest    3b418d7b466a   5 weeks ago      77.8MB
ubuntu       20.04     88bd68917189   7 weeks ago      72.8MB
PS C:\Users\mikel\repo\DUNE>

```

Reply:
I believe release tagging is fixed in f6936b2. But I think a test still needs to be added.

Reply:
I believe release tagging is fixed in f6936b2. But tests needed to be added.

Reply:
Significant tests were added.